doc: add Hex docs documentation

randycoulman opened the pull request:
Closes #4

TODO:
- [x] Update once #55 and #56 have either been merged or rejected.
- [x] Add a better version of the logo (to be provided by ConfigCat's designer via @laliconfigcat).
- [x] Use `@typep` and `@opaque` to undo some of the type-weakening changes made here.

Uses ex_doc to add Elixir documentation to the public modules and functions.  I copied some of the documentation from the README, from ConfigCat's website, and from @igorescobar's work on https://github.com/configcat/docs/pull/21.

I'm not sure of the documentation about variation id's, so comments on that would be most welcome.  That feature is not yet documented elsewhere, so I tried to make up something that sounded reasonable.

Includes README.md in the generated documentation. To make this work, I had to:
- Rename `media` to `assets`
- Use markdown links rather than `a` tags
- Tweak some back-quoted identifiers so that they become automatic links in the API docs.

In order to avoid reference private modules, I made a few changes to some of the typespecs.

To explore the documentation locally, you can run `mix docs` and then open `doc/index.html` in your browser.

Commit:
chore: add ex_doc dependency

Commit:
feat: add extra attributes for hexdocs site

Commit:
doc: Omit private modules from Hex docs

This results in a number of warnings about referencing private types; I'll address those in a later commit.

Commit:
doc: omit private functions and types from docs

Commit:
doc: add README to hex docs

In order for this to work, I had to change some things:

- Rename `media` to `assets`
- Use markdown links rather than `a` tags
- Tweak some back-quoted identifiers so that they become automatic links into the API docs.

Commit:
doc: add logo

Commit:
doc: add docs for ConfigCache

I decided to make the `Config` module public in the docs so we can reference its (public) types.

I also inlined the `result` type because it was adding an extra layer of indirection in the docs and was only used one other time in a private module.

Commit:
doc: add documentation to Config

Commit:
doc: add Hex docs for User

Commit:
doc: add docs for CachePolicy

I had to tweak some of the typedefs to avoid referencing internal, non-documented types.

Commit:
doc: add docs for ConfigCat

Commit:
doc: include proper logo

randycoulman commented:
I just remembered that Elixir has private and opaque types. Before I merge this, I'm going to revert some of the type changes I made in this PR and use those features instead.

igorescobar commented on README.md:
```diff
@@ -1,18 +1,20 @@
 ConfigCat SDK for Elixir provides easy integration for your application to ConfigCat.
 
-ConfigCat is a feature flag and configuration management service that lets you separate releases from deployments. You can turn your features ON/OFF using <a href="http://app.configcat.com" target="_blank">ConfigCat Dashboard</a> even after they are deployed. ConfigCat lets you target specific groups of users based on region, email or any other custom user attribute.
+ConfigCat is a feature flag and configuration management service that lets you separate releases from deployments. You can turn your features ON/OFF using [ConfigCat Dashboard](http://app.configcat.com) even after they are deployed. ConfigCat lets you target specific groups of users based on region, email or any other custom user attribute.
```
I believe this change was intentional. Like this, the link opens on the same tab by default and I think they wanted to use `target="_blank"` because of its behavior.

randycoulman replied:
I wondered about that, but these look really bad in HexDocs, because it doesn't handle literal `a` tags.  We could make a copy of the README for use in the HexDocs, but that's an additional maintenance burden.

This same consideration arises with your comment below about `media` vs `assets`. We can put the image in a `media` folder and specify that ex_doc should copy it to the `assets` folder in the docs (we have no choice about the name there), but then the reference to it in the README then only works in once place or the other.

@laliconfigcat What do you want to do here? Do we want to maintain consistency with the other SDKs at the expense of having to maintain a second almost-identical document? Or is it OK to have a couple of relatively minor differences in order to work better with Elixir's standard tools here?

laliconfigcat replied:
It's fine for us without the <a> tags and also the media -> assets folder

igorescobar commented on README.md:
```diff
@@ -1,18 +1,20 @@
+ConfigCat is a feature flag and configuration management service that lets you separate releases from deployments. You can turn your features ON/OFF using [ConfigCat Dashboard](http://app.configcat.com) even after they are deployed. ConfigCat lets you target specific groups of users based on region, email or any other custom user attribute.
 
-ConfigCat is a <a href="https://configcat.com" target="_blank">hosted feature flag service</a>. Manage feature toggles across frontend, backend, mobile, desktop apps. <a href="https://configcat.com" target="_blank">Alternative to LaunchDarkly</a>. Management app + feature flag SDKs.
+ConfigCat is a [hosted feature flag service](http://configcat.com). Manage feature toggles across frontend, backend, mobile, desktop apps. [Alternative to LaunchDarkly](http://configcat.com). Management app + feature flag SDKs.
```
Same here

igorescobar commented on README.md:
```diff
@@ -22,8 +24,10 @@ def deps do
-### 2. Go to <a href="https://app.configcat.com/sdkkey" target="_blank">Connect your application</a> tab to get your *SDK Key*:
-![SDK-KEY](./media/readme01.png  "SDK-KEY")
+
+### 2. Go to [Connect your application](https://app.configcat.com/sdkkey) tab to get your _SDK Key_:
```
And here 😅

igorescobar commented on README.md:
```diff
@@ -22,8 +24,10 @@ def deps do
+
+### 2. Go to [Connect your application](https://app.configcat.com/sdkkey) tab to get your _SDK Key_:
+
+![SDK-KEY](assets/readme01.png "SDK-KEY")
```
They call this folder `media/` on all SDKs. So I named the same intentionally :)

randycoulman replied:
See discussion above.

igorescobar commented:
So it's all good 🪂

Commit:
refactor: minor type cleanups

- Use `@opaque` to restore the proper definition of `CachePolicy.t()`

- Extract a few more reusable types

randycoulman commented:
@igorescobar Ready for final review now. I ended up not using `typep` anywhere because that hides the types from other modules in the library which I didn't want. But I was able to restore the definition of `CachePolicy.t()` by using `@opaque`.

igorescobar commented:
🎉

igorescobar commented:
@randycoulman I was able to check the docs locally and it looks good to me. I will go ahead and merge this PR 👍
